src/probe_modules/module_dns.c

```diff
@@ -0,0 +1,1043 @@
+            }
+
+            // Did we parse OK?
+            fs_add_uint64(fs, "dns_parse_err", err); 
+        }
+    
+        // Now the raw stuff.
+        fs_add_binary(fs, "raw_data", (udp_len - sizeof(struct udphdr)), (void*) &udp_hdr[1], 0);
+        //fs_add_binary(fs, "raw_data", len, (void*)packet, 0);
+   
+        return;
+    
+    } else if (ip_hdr->ip_p == IPPROTO_ICMP) {
+        struct icmp *icmp = (struct icmp*) ((char *) ip_hdr + 4*ip_hdr->ip_hl);
+        struct ip *ip_inner = (struct ip*) ((char *) icmp + ICMP_UNREACH_HEADER_SIZE);
+        
+        // This is the packet we sent
+        struct udphdr *udp_hdr = (struct udphdr *) ((char*) ip_inner + 4*ip_inner->ip_hl);
+        
+        uint16_t udp_len = ntohs(udp_hdr->uh_ulen);
+
+        // High level info    
+        fs_add_string(fs, "classification", (char*) "dns", 0);
+        fs_add_uint64(fs, "success", 0);
+        fs_add_uint64(fs, "app_success", 0);
+        
+        // UDP info
+        fs_add_uint64(fs, "udp_sport", ntohs(udp_hdr->uh_sport));
+        fs_add_uint64(fs, "udp_dport", ntohs(udp_hdr->uh_dport));
+        fs_add_uint64(fs, "udp_len", udp_len);
+        
+        // ICMP info
+        // XXX This is legacy. not well tested.
+        fs_add_string(fs, "icmp_responder", make_ip_str(ip_hdr->ip_src.s_addr), 1);
+        fs_add_uint64(fs, "icmp_type", icmp->icmp_type);
+        fs_add_uint64(fs, "icmp_code", icmp->icmp_code);
+        if (icmp->icmp_code <= ICMP_UNREACH_PRECEDENCE_CUTOFF) {
+            fs_add_string(fs, "icmp_unreach_str",
+                (char *) udp_unreach_strings[icmp->icmp_code], 0);
+        } else {
+            fs_add_string(fs, "icmp_unreach_str", (char *) "unknown", 0);
+        }
+        
+        // DNS header
+        fs_add_null(fs, "dns_id"); 
+        fs_add_null(fs, "dns_rd"); 
+        fs_add_null(fs, "dns_tc"); 
+        fs_add_null(fs, "dns_aa"); 
+        fs_add_null(fs, "dns_opcode"); 
+        fs_add_null(fs, "dns_qr"); 
+        fs_add_null(fs, "dns_rcode"); 
+        fs_add_null(fs, "dns_cd"); 
+        fs_add_null(fs, "dns_ad"); 
+        fs_add_null(fs, "dns_z"); 
+        fs_add_null(fs, "dns_ra"); 
+        fs_add_null(fs, "dns_qdcount"); 
+        fs_add_null(fs, "dns_ancount"); 
+        fs_add_null(fs, "dns_nscount"); 
+        fs_add_null(fs, "dns_arcount"); 
+
+        fs_add_null(fs, "dns_questions");
+        fs_add_null(fs, "dns_answers");
+        fs_add_null(fs, "dns_authorities");
+        fs_add_null(fs, "dns_additionals");
+
+        fs_add_uint64(fs, "dns_unconsumed_bytes", 0); 
+        fs_add_uint64(fs, "dns_parse_err", 1); 
+        fs_add_binary(fs, "raw_data", len, (char*)packet, 0);
+        
+        return;
+
+    } else {
+        // This should not happen. Both the pcap filter and validate packet prevent this.
+        assert(0);
+        return;
+    }
+}
+
+static fielddef_t fields[] = {
+    {.name = "classification", .type="string", .desc = "packet protocol"},
```

Comment:
Fun fact: this syntax is actually a GNU extension. Clang also supports it, so I'm all for using it.